### ISSUE TITLE
ICE: fail-if-zero on unexpected type: core::simd::u32x4

### ISSUE THREAD

**Issue by msiemens** [rust-lang/rust#19917]
Code:
```rust
use std::simd::u32x4;

fn function(n: u32, key: u32x4) -> u32x4 {
    key % u32x4(n, n, n, n)
}
```

Error:
```
error: internal compiler error: fail-if-zero on unexpected type: core::simd::u32x4
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
note: run with `RUST_BACKTRACE=1` for a backtrace
task 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-linux/build/src/libsyntax/diagnostic.rs:180

stack backtrace:
   1:     0x7fa6a4006e50 - rt::backtrace::imp::write::hf87c140cfcd95fbaBly
   2:     0x7fa6a400a0a0 - failure::on_fail::h343c4b97464b676eZNy
   3:     0x7fa6a3c5cfa0 - unwind::begin_unwind_inner::hb4b9831d2f5e86936Qc
   4:     0x7fa6a16c6150 - unwind::begin_unwind::h9628242264111950763
   5:     0x7fa6a16c6a00 - diagnostic::Handler::bug::h15005c99f56ee693ubG
   6:     0x7fa6a2279390 - session::Session::bug::hceed30836fdfb2a6aEE
   7:     0x7fa6a2cd6980 - trans::base::fail_if_zero_or_overflows::h28c84f6ec45170d2Wvt
   8:     0x7fa6a2cd5630 - trans::expr::trans_eager_binop::hafeec29ec0e70d88xqk
   9:     0x7fa6a2cca7f0 - trans::expr::trans_binary::h2aecf3dc843f5d7fGzk
  10:     0x7fa6a2cb4770 - trans::expr::trans_unadjusted::h55e0732b75e1ef2bpSi
  11:     0x7fa6a2c6d000 - trans::expr::trans_into::hdb5c2dee155d90ablZh
  12:     0x7fa6a2c6d760 - trans::controlflow::trans_block::hcc615213ff54c86fu2d
  13:     0x7fa6a2d1d300 - trans::base::trans_closure::hbffb384cd75712bc3Lu
  14:     0x7fa6a2c601d0 - trans::base::trans_fn::h62cf29b3a9287cf5UXu
  15:     0x7fa6a2c5a610 - trans::base::trans_item::hbec23991f40c487eDjv
  16:     0x7fa6a2d21d50 - trans::base::trans_mod::hc3a8158957ec37b18ov
  17:     0x7fa6a2d25be0 - trans::base::trans_crate::h402323672a693eebxhw
  18:     0x7fa6a4446300 - driver::phase_4_translate_to_llvm::h7e410f8bfecf8edcNCa
  19:     0x7fa6a442b100 - driver::compile_input::h022976239e8f5a3cpba
  20:     0x7fa6a44c7d80 - run_compiler::h2b243518ea560709z4b
  21:     0x7fa6a44c7c70 - run::closure.21242
  22:     0x7fa6a44d9670 - task::TaskBuilder::try_future::closure.22690
  23:     0x7fa6a3fdf410 - task::TaskBuilder::spawn_internal::closure.31179
  24:     0x7fa6a3c5ad70 - task::Task::spawn::closure.5765
  25:     0x7fa6a3cb8280 - rust_try_inner
  26:     0x7fa6a3cb8270 - rust_try
  27:     0x7fa6a3c5ae50 - unwind::try::ha36bc8815a761f13pFc
  28:     0x7fa6a3c5ac10 - task::Task::run::h24525a73fb4c9bb3fOb
  29:     0x7fa6a3c5a800 - task::Task::spawn::closure.5741
  30:     0x7fa6a3c5c250 - thread::thread_start::ha943fdda31374bc3i6b
  31:     0x7fa69eb66460 - start_thread
  32:     0x7fa6a392f749 - clone
  33:                0x0 - <unknown>
```

```
$ rustc --version
rustc 0.13.0-dev (ef0bc464a 2014-12-15 11:17:44 +0000)
```

**Comment by msiemens** [comment]
This is fixed in nightly, propably due to #23704.

**Comment by msiemens** [comment]
Altough this propably should need a test, too
